src/pull-retry.ts

```diff
@@ -0,0 +1,94 @@
+/*!
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {StatusObject, status} from 'grpc';
+
+export interface PullRetryOptions {
+  maxTimeout: number;
+}
+
+/*!
+ * retryable status codes
+ */
+const RETRY_CODES: status[] = [
+  status.OK,
+  status.CANCELLED,
+  status.UNKNOWN,
+  status.DEADLINE_EXCEEDED,
+  status.RESOURCE_EXHAUSTED,
+  status.ABORTED,
+  status.INTERNAL,
+  status.UNAVAILABLE,
+  status.DATA_LOSS,
+];
+
+/**
+ * Used to track pull requests and determine if additional requests should be
+ * made, etc.
+ *
+ * @class
+ * @private
+ */
+export class PullRetry {
+  private failures = 0;
+  private lastSuccess = Date.now();
+  private maxTimeout: number;
+  constructor(options = {} as PullRetryOptions) {
+    this.maxTimeout = options.maxTimeout;
+  }
+  /**
+   * Generates a timeout that can be used for applying a backoff based on the
+   * current number of failed requests.
+   *
+   * @private
+   * @returns {number}
+   */
+  createTimeout(): number {
+    if (this.failures === 0) {
+      return 0;
+    }
+    return Math.pow(2, this.failures) * 1000 + Math.floor(Math.random() * 1000);
```

Comment:
I might be tempted to add a comment:

1. describing the algorithm in brief (backoff by failures^2 seconds + a random 1-second skew?)
2. what's the justification for these #s, curious if we use them elsewhere?

Reply:
I pulled these from an older version of the PubSub retry logic. Originally I believe this came from [`retry-request`](https://www.npmjs.com/package/retry-request), which is the library we use for handling retries in the `nodejs-common` and `nodejs-common-grpc` modules. I would probably ask @stephenplusplus to explain the numbers though.

Reply:
[Boom!](https://cloud.google.com/iot/docs/how-tos/exponential-backoff)

Reply:
(A similar document seems to exist for each API, but in general, that's the approach that was found ~4/5 years ago when we wrote that little bit)

Reply:
awesome, perhaps just add a comment linking to this document; so that future humans find it, and don't think the approach was pulled out of thin air.

Reply:
Done, add a `@see` tag to the JSDoc.